test/test_utilities.cpp

```diff
@@ -116,5 +104,14 @@ void ValidateTensorData(Tensor& T1, Tensor& T2, float tol) {
   }
 }
 
+template <>
+bool eq(float arg0, float arg1) {
+  if (arg0 == 0 && arg1 == 0) {
+    return true;
+  } else {
+    return (abs(arg0 - arg1) / max(abs(arg0), abs(arg1)) <= 0.001);
```

Comment:
Should I add a cout here which would print the values in case of false(failing test) rather than enabling the log levels?

Reply:
I am not able to see the AssertTensorEquals() function on the git. Another way would be to add a switch(T) inside the AssertTensorEquals() and let GTEST do all the heavy lifting of validating float, int outputs and generating clean failure logs. @sayantan-nervana  @kanvi-nervana what do you think?

template<T>
void AssertTensorEquals(Tensor& T1, Tensor& T2) {
  ASSERT_EQ(T1.shape(), T2.shape());
  auto T_size = T1.flat<T>().size();
  auto T1_data = T1.flat<T>().data();
  auto T2_data = T2.flat<T>().data();
  for (int k = 0; k < T_size; k++) {
    auto a = T1_data[k];
    auto b = T2_data[k];
    switch(T){
       case float:
         EXPECT_FLOAT_EQ(a, b);
         break;
       case int:
         EXPECT_EQ(a, b);
         break;
       case ....
     }
  }
}

Reply:
Need to validate if switch works like this on template : switch(T)

Reply:
Validated from @sayantan-nervana .. tf datatypes DT_FLOAT, DT_INT32 are defined as enum .. so this might work.

Reply:
it would be:
```template<int T>```
or with better types:
```template<tensorflow_enum_type T>```
But then, why template it (except maybe to get _slightly_ faster runtimes at the expense of _slightly_ slower compile times. We might as well pass the tensorflow datatype as a param to AssertTensorEquals.

Not sure this line will not work: T1.flat<T>().data();
current usage: ```T1.flat<float>().data()``` . ```flat``` seems to take in real data types, so tensorflow enum datatypes may not work.

Reply:
Following up on discussion from slack:

```

template <DataType T>  //an enum/int
void AssertTensorEquals(Tensor& T1, Tensor& T2) {
  ASSERT_EQ(T1.shape(), T2.shape());
  auto T_size = T1.flat<EnumToDataType<T>::Type>().size();
  auto T1_data = T1.flat<EnumToDataType<T>::Type>().data();
  auto T2_data = T2.flat<EnumToDataType<T>::Type>().data();
  for (int k = 0; k < T_size; k++) {
    auto a = T1_data[k];
    auto b = T2_data[k];
    switch (T){
      case DT_FLOAT:
      case DT_DOUBLE:
        EXPECT_FLOAT_EQ(a, b);
        break;
      case DT_INT32:
      case DT_OTHER_INTS:
        EXPECT_EQ(a, b);
    }
  }
}
```

Reply:
Thanks .. This is super! 👍 Addresses all our concerns.

Reply:
This is getting into a template hell with difficult-to-parse compilation errors, debugging which would need time/effort. So suggest we not use BKM and switch to second best BKM discussed earlier.

Reply:
Thanks for trying! 👍